### PR TITLE
make the logger and config singleton.

### PR DESCRIPTION
**Pull request by andyliuliming**

### CODE REVIEW

File: src/bosh_azure_cpi/lib/cloud/azure/models/config.rb [hunk 1]
```diff
@@ -106,31 +107,43 @@ class RegistryConfig
...
 
     def to_h
-      @config
+      @config_hash
     end
   end
 
   class Config
+    include Singleton
+    include Helpers
```
Comment (gossion) [inline src/bosh_azure_cpi/lib/cloud/azure/models/config.rb]: what is the purpose of  `include Helpers`?

File: src/bosh_azure_cpi/lib/cloud/azure/cloud.rb [hunk 2]
```diff
@@ -14,21 +13,15 @@ class Cloud < Bosh::Cloud
...
-      options_dup = options.dup.freeze
-
-      @logger = Bosh::Clouds::Config.logger
-
-      @config = Bosh::AzureCloud::ConfigFactory.build(options_dup)
-
-      request_id = options_dup['azure']['request_id']
-      Bosh::AzureCloud::CPILogger.set_request_id(request_id) if request_id
+    def initialize
+      request_id = config.azure.request_id
```
Comment (gossion) [inline src/bosh_azure_cpi/lib/cloud/azure/cloud.rb]: where is the code to initialize `Bosh::AzureCloud::Config`?

File: src/bosh_azure_cpi/lib/cloud/azure/telemetry/telemetry_manager.rb [hunk 3]
```diff
@@ -10,7 +10,7 @@ class TelemetryManager
 
     def initialize(azure_config)
       @azure_config = azure_config
-      @logger = Bosh::Cpi::Logger.new(CPI_TELEMETRY_LOG_FILE)
+      Bosh::AzureCloud::CPILogger.instance.logger = Bosh::Cpi::Logger.new(CPI_TELEMETRY_LOG_FILE)
```
Comment (gossion) [inline src/bosh_azure_cpi/lib/cloud/azure/telemetry/telemetry_manager.rb]: the logger is changed.. I really doubt if the logs are written to bosh director (STDERR).

Comment (andyliuliming) [reply src/bosh_azure_cpi/lib/cloud/azure/telemetry/telemetry_manager.rb]: good catch.

**Comment by gossion** [comment]
I did not see the necessity to make this change, the improvement is very little, and even introduce bugs. Furthermore,  we need to waste a lot of time write / review the code. please don't do such change again.

**Comment by andyliuliming** [comment]
> I did not see the necessity to make this change, the improvement is very little, and even introduce bugs. Furthermore, we need to waste a lot of time write / review the code. please don't do such change again.

please let me explain why I need this change:
1. for doing the vmss implementation, I will provide one way to run our cpi as one rpc service, for cleaning up the disk and handle the batching requests. based on that, the config and the logger should be the singleton.
2. and to seperate the "logger" in our cpi and the bosh cli has its value, our cpi should not be dependent on the "Bosh::Clouds::Config." component, because that the bosh component.
3. and if you are too busy to review this, feel free go reject the review.

**Comment by gossion** [comment]
1. for the logger part, in your rpc service maybe you can't use the same logger with main process: if there are logs posted to its STDERR after create_vm is done, the main process can't quit (it is possible, but I did not try).
2. for "Bosh::Clouds::Config.", other CPI like aws also uses the same mechanism. it is not needed to decouple this dependency.

**Comment by gossion** [comment]
I will continue the review, but if the reason is more consolidated, I will be happier.

**Comment by andyliuliming** [comment]
> 1. for the logger part, in your rpc service maybe you can't use the same logger with main process: if there are logs posted to its STDERR after create_vm is done, the main process can't quit (it is possible, but I did not try).
Yes, that's why the logger writtable. 


> 2. for "Bosh::Clouds::Config.", other CPI like aws also uses the same mechanism. it is not needed to decouple this dependency.

we are azure :)

**Comment by andyliuliming** [comment]
> I will continue the review, but if the reason is more consolidated, I will be happier.

will fill the reason form next time. thanks.

File: src/bosh_azure_cpi/lib/cloud/azure/cloud.rb (hunk 2 above)
Comment (andyliuliming) [reply src/bosh_azure_cpi/lib/cloud/azure/cloud.rb]: added.

File: src/bosh_azure_cpi/lib/cloud/azure/models/config.rb (hunk 1 above)
Comment (andyliuliming) [reply src/bosh_azure_cpi/lib/cloud/azure/models/config.rb]: removed the extra include.

File: src/bosh_azure_cpi/lib/cloud/azure/cloud.rb (hunk 2 above)
Comment (andyliuliming) [reply src/bosh_azure_cpi/lib/cloud/azure/cloud.rb]: added.

**Pull request merged**